Task

Train SVM classifier using sklearn digits dataset (i.e. from sklearn.datasets import load_digits) and then,

1. Measure accuracy of your model using different kernels such as rbf and linear.

2. Tune your model further using regularization and gamma parameters and try to come up with highest accurancy score

3. Use 80% of samples as training data size

In [4]:
import pandas as pd
from sklearn.datasets import load_digits
digits = load_digits()

In [5]:
dir(digits)

['DESCR', 'data', 'feature_names', 'frame', 'images', 'target', 'target_names']

In [6]:
digits.target

array([0, 1, 2, ..., 8, 9, 8])

In [8]:
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [9]:
# create df for our datasets
df = pd.DataFrame(digits.data, digits.target)
df.head()

0    1    2     3     4     5    6    7    8    9   ...   54   55   56  \
0  0.0  0.0  5.0  13.0   9.0   1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1  0.0  0.0  0.0  12.0  13.0   5.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2  0.0  0.0  0.0   4.0  15.0  12.0  0.0  0.0  0.0  0.0  ...  5.0  0.0  0.0   
3  0.0  0.0  7.0  15.0  13.0   1.0  0.0  0.0  0.0  8.0  ...  9.0  0.0  0.0   
4  0.0  0.0  0.0   1.0  11.0   0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

    57   58    59    60    61   62   63  
0  0.0  6.0  13.0  10.0   0.0  0.0  0.0  
1  0.0  0.0  11.0  16.0  10.0  0.0  0.0  
2  0.0  0.0   3.0  11.0  16.0  9.0  0.0  
3  0.0  7.0  13.0  13.0   9.0  0.0  0.0  
4  0.0  0.0   2.0  16.0   4.0  0.0  0.0  

[5 rows x 64 columns]

In [11]:
# append target to the dataframe
df["target"] = digits.target
df.head(15)

0    1     2     3     4     5     6    7    8    9  ...   55   56   57  \
0  0.0  0.0   5.0  13.0   9.0   1.0   0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1  0.0  0.0   0.0  12.0  13.0   5.0   0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2  0.0  0.0   0.0   4.0  15.0  12.0   0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3  0.0  0.0   7.0  15.0  13.0   1.0   0.0  0.0  0.0  8.0  ...  0.0  0.0  0.0   
4  0.0  0.0   0.0   1.0  11.0   0.0   0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
5  0.0  0.0  12.0  10.0   0.0   0.0   0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
6  0.0  0.0   0.0  12.0  13.0   0.0   0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
7  0.0  0.0   7.0   8.0  13.0  16.0  15.0  1.0  0.0  0.0  ...  0.0  0.0  0.0   
8  0.0  0.0   9.0  14.0   8.0   1.0   0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
9  0.0  0.0  11.0  12.0   0.0   0.0   0.0  0.0  0.0  2.0  ...  0.0  0.0  0.0   
0  0.0  0.0   1.0   9.0  15.0  11.0   0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1  0.0  0.0   0.0   0.0  14.0  13.0   1.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2  0.0  0.0   5.0  12.0   1.0   0.0   0.0  0.0  0.0  0.0  ...  2.0  0.0  0.0   
3  0.0  2.0   9.0  15.0  14.0   9.0   3.0  0.0  0.0  4.0  ...  0.0  0.0  2.0   
4  0.0  0.0   0.0   8.0  15.0   1.0   0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

     58    59    60    61    62   63  target  
0   6.0  13.0  10.0   0.0   0.0  0.0       0  
1   0.0  11.0  16.0  10.0   0.0  0.0       1  
2   0.0   3.0  11.0  16.0   9.0  0.0       2  
3   7.0  13.0  13.0   9.0   0.0  0.0       3  
4   0.0   2.0  16.0   4.0   0.0  0.0       4  
5   9.0  16.0  16.0  10.0   0.0  0.0       5  
6   1.0   9.0  15.0  11.0   3.0  0.0       6  
7  13.0   5.0   0.0   0.0   0.0  0.0       7  
8  11.0  16.0  15.0  11.0   1.0  0.0       8  
9   9.0  12.0  13.0   3.0   0.0  0.0       9  
0   1.0  10.0  13.0   3.0   0.0  0.0       0  
1   0.0   1.0  13.0  16.0   1.0  0.0       1  
2   3.0  11.0   8.0  13.0  12.0  4.0       2  
3  12.0  12.0  13.0  11.0   0.0  0.0       3  
4   0.0  10.0  15.0   4.0   0.0  0.0       4  

[15 rows x 65 columns]

In [14]:
# define X and y
X = df.drop("target", axis= "columns")
y = df.target

In [15]:
X.head()

0    1    2     3     4     5    6    7    8    9   ...   54   55   56  \
0  0.0  0.0  5.0  13.0   9.0   1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1  0.0  0.0  0.0  12.0  13.0   5.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2  0.0  0.0  0.0   4.0  15.0  12.0  0.0  0.0  0.0  0.0  ...  5.0  0.0  0.0   
3  0.0  0.0  7.0  15.0  13.0   1.0  0.0  0.0  0.0  8.0  ...  9.0  0.0  0.0   
4  0.0  0.0  0.0   1.0  11.0   0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

    57   58    59    60    61   62   63  
0  0.0  6.0  13.0  10.0   0.0  0.0  0.0  
1  0.0  0.0  11.0  16.0  10.0  0.0  0.0  
2  0.0  0.0   3.0  11.0  16.0  9.0  0.0  
3  0.0  7.0  13.0  13.0   9.0  0.0  0.0  
4  0.0  0.0   2.0  16.0   4.0  0.0  0.0  

[5 rows x 64 columns]

In [16]:
y.head()

0    0
1    1
2    2
3    3
4    4
Name: target, dtype: int32

In [18]:
# model training
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size= 0.2)

In [20]:
len(X_train)

1437

In [21]:
len(X_test)

360

Measure accuracy of your model using different kernels such as rbf and linear.

In [22]:
from sklearn.svm import SVC


In [25]:
rbf = SVC(kernel= "rbf")
rbf

SVC()

In [29]:
rbf.fit(X_train, y_train)


SVC()

In [31]:
rbf.score(X_test, y_test)

0.9944444444444445

In [34]:
digits.target[607]

3

In [38]:
rbf.predict(digits.data[[607]])

array([3])

In [40]:
y_predicted = rbf.predict(X_test)
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_predicted)
cm

array([[32,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 33,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0, 38,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 38,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0, 41,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0, 36,  0,  0,  0,  0],
       [ 0,  1,  0,  0,  0,  0, 35,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0, 32,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0, 29,  1],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0, 44]], dtype=int64)

using linear kernel

In [42]:
linear = SVC(kernel= "linear")
linear.fit(X_train, y_train)

SVC(kernel='linear')

In [43]:
linear.score(X_test, y_test)

0.9833333333333333

In [45]:
linear.predict(digits.data[[607]])

array([3])

In [46]:
y_predicted = linear.predict(X_test)
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_predicted)
cm

array([[32,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 33,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0, 38,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 38,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0, 41,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0, 36,  0,  0,  0,  0],
       [ 0,  1,  0,  0,  0,  0, 35,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0, 32,  0,  0],
       [ 0,  0,  0,  0,  1,  2,  0,  0, 27,  0],
       [ 0,  0,  0,  1,  0,  1,  0,  0,  0, 42]], dtype=int64)

 further tuning with regularization(C) and gamma

In [60]:
# using linear anf gamma auto
print("***** USING linear AND GAMMA=AUTO(set C: 1.0) ******")
linearga = SVC(C= 1.0, kernel= "linear", gamma= "auto")
linearga.fit(X_train, y_train)
print("Score for linear and gamma=auto: ", linearga.score(X_test, y_test))
print("predict digits.data[[607]]: ", linearga.predict(digits.data[[607]]))

y_predicted = linearga.predict(X_test)
print("y_predicted: \n", y_predicted)


cm = confusion_matrix(y_test, y_predicted)
print("CONFUISON MATRIX: \n", cm)

***** USING linear AND GAMMA=AUTO(set C: 1.0) ******
Score for linear and gamma=auto:  0.9833333333333333
predict digits.data[[607]]:  [3]
y_predicted: 
 [3 9 4 8 4 8 9 1 3 9 0 2 8 8 7 0 9 2 7 4 2 6 4 3 5 8 4 2 1 8 7 7 8 7 0 6 1
 5 0 5 8 6 7 4 2 2 9 2 1 6 5 7 3 4 1 0 0 4 5 3 2 0 9 5 4 4 9 8 1 2 8 1 8 6
 5 5 3 0 1 3 1 3 8 1 7 8 1 8 3 0 4 0 6 2 4 3 2 6 6 5 4 2 2 4 6 6 9 5 1 0 6
 4 3 0 5 7 5 6 8 4 6 4 7 6 7 4 2 9 2 4 9 8 9 7 4 0 3 9 3 9 0 6 2 9 5 3 3 9
 9 1 5 8 6 5 3 4 4 9 7 5 1 5 9 0 8 6 9 9 6 3 2 9 0 9 2 3 7 6 4 1 1 7 9 4 7
 0 8 4 9 8 2 2 5 3 9 4 9 5 7 1 1 7 8 6 1 1 6 9 5 3 1 2 2 4 6 1 2 8 1 7 3 0
 3 6 3 5 1 3 0 6 3 2 1 5 1 0 4 8 5 5 4 3 7 4 7 9 2 9 9 7 6 2 7 0 2 2 3 7 0
 5 5 6 6 5 5 5 7 3 3 0 3 3 4 5 5 8 6 7 9 6 7 4 4 3 8 2 1 9 6 3 6 7 3 4 2 8
 5 0 3 0 2 2 0 2 6 5 5 2 0 2 1 5 4 3 1 1 1 3 9 6 9 9 5 4 4 4 7 7 9 1 9 5 1
 3 7 0 4 1 9 2 4 2 0 8 5 0 6 9 3 9 6 7 0 5 9 4 0 9 2 4]
CONFUISON MATRIX: 
 [[32  0  0  0  0  0  0  0  0  0]
 [ 0 33  0  0  0  0  0  0  0  0]
 [ 0  0 38  0  0  0  0  0  0 

In [61]:
# using linear and gamma scale
print("***** USING linear AND GAMMA=SCALE(set C: 1.0) ******")
lineargas = SVC(C= 1.0, kernel= "linear", gamma= "scale")
lineargas.fit(X_train,y_train)
print("Score for rbf and gamma=scale: ", lineargas.score(X_test, y_test))
print("predict digits.data[[607]]: ", lineargas.predict(digits.data[[607]]))

y_predicted = lineargas.predict(X_test)
print("y_predicted: \n", y_predicted)


cm = confusion_matrix(y_test, y_predicted)
print("CONFUISON MATRIX: \n", cm)

***** USING linear AND GAMMA=SCALE(set C: 1.0) ******
Score for rbf and gamma=scale:  0.9833333333333333
predict digits.data[[607]]:  [3]
y_predicted: 
 [3 9 4 8 4 8 9 1 3 9 0 2 8 8 7 0 9 2 7 4 2 6 4 3 5 8 4 2 1 8 7 7 8 7 0 6 1
 5 0 5 8 6 7 4 2 2 9 2 1 6 5 7 3 4 1 0 0 4 5 3 2 0 9 5 4 4 9 8 1 2 8 1 8 6
 5 5 3 0 1 3 1 3 8 1 7 8 1 8 3 0 4 0 6 2 4 3 2 6 6 5 4 2 2 4 6 6 9 5 1 0 6
 4 3 0 5 7 5 6 8 4 6 4 7 6 7 4 2 9 2 4 9 8 9 7 4 0 3 9 3 9 0 6 2 9 5 3 3 9
 9 1 5 8 6 5 3 4 4 9 7 5 1 5 9 0 8 6 9 9 6 3 2 9 0 9 2 3 7 6 4 1 1 7 9 4 7
 0 8 4 9 8 2 2 5 3 9 4 9 5 7 1 1 7 8 6 1 1 6 9 5 3 1 2 2 4 6 1 2 8 1 7 3 0
 3 6 3 5 1 3 0 6 3 2 1 5 1 0 4 8 5 5 4 3 7 4 7 9 2 9 9 7 6 2 7 0 2 2 3 7 0
 5 5 6 6 5 5 5 7 3 3 0 3 3 4 5 5 8 6 7 9 6 7 4 4 3 8 2 1 9 6 3 6 7 3 4 2 8
 5 0 3 0 2 2 0 2 6 5 5 2 0 2 1 5 4 3 1 1 1 3 9 6 9 9 5 4 4 4 7 7 9 1 9 5 1
 3 7 0 4 1 9 2 4 2 0 8 5 0 6 9 3 9 6 7 0 5 9 4 0 9 2 4]
CONFUISON MATRIX: 
 [[32  0  0  0  0  0  0  0  0  0]
 [ 0 33  0  0  0  0  0  0  0  0]
 [ 0  0 38  0  0  0  0  0  0  

no observable change

In [64]:
# using rbf and gamma auto
print("***** USING rbf AND GAMMA=AUTO(set C: 1.0) ******")
rbfga = SVC(C= 1.0, kernel= "rbf", gamma= "auto")
rbfga.fit(X_train,y_train)
print("Score for rbf and gamma=auto: ", rbfga.score(X_test, y_test))
print("predict digits.data[[607]]: ", rbfga.predict(digits.data[[607]]))

y_predicted = rbfga.predict(X_test)
print("y_predicted: \n", y_predicted)


cm = confusion_matrix(y_test, y_predicted)
print("CONFUISON MATRIX: \n", cm)

***** USING rbf AND GAMMA=AUTO(set C: 1.0) ******
Score for rbf and gamma=auto:  0.5916666666666667
predict digits.data[[607]]:  [3]
y_predicted: 
 [3 8 8 8 4 8 8 1 3 8 0 8 8 8 8 0 8 2 7 8 2 6 4 3 5 8 8 8 1 8 8 8 8 7 8 8 1
 8 0 5 8 6 8 4 8 8 8 2 8 6 5 7 3 8 8 0 8 4 5 8 8 0 8 5 4 8 8 8 1 8 8 8 8 6
 5 5 8 8 1 8 8 3 8 1 7 8 1 8 3 0 8 8 6 8 8 3 8 6 6 5 8 8 8 8 6 8 8 5 1 0 8
 4 3 8 5 7 5 6 8 4 6 8 7 8 7 8 8 8 8 8 9 8 8 8 4 0 3 9 3 8 8 6 8 8 5 3 3 8
 8 8 5 8 6 8 3 8 4 8 7 5 1 5 9 0 8 6 9 9 8 8 8 8 0 8 8 3 8 6 8 8 1 8 8 4 7
 8 8 8 9 8 8 2 5 3 8 8 5 5 7 1 8 7 8 8 1 8 6 8 8 3 1 8 8 8 6 1 8 8 8 7 3 0
 3 8 3 5 1 8 8 6 3 8 8 5 1 8 8 8 5 5 4 3 7 4 7 8 2 8 8 7 8 2 7 8 8 8 3 7 8
 5 5 6 6 5 5 5 7 3 3 8 3 3 8 5 5 8 8 7 9 6 7 4 4 3 8 2 8 8 6 3 8 8 8 4 8 8
 5 8 3 0 8 8 8 8 8 5 5 8 0 8 1 5 8 3 8 8 1 3 9 6 8 8 5 8 8 8 7 7 8 1 8 5 8
 3 7 0 8 8 8 8 4 8 0 8 8 0 8 8 3 8 8 7 0 5 8 8 0 8 2 8]
CONFUISON MATRIX: 
 [[18  0  0  0  0  0  0  0 14  0]
 [ 0 19  0  0  0  0  0  0 14  0]
 [ 0  0  8  0  0  0  0  0 30  0]
 [

In [65]:
# using rbf and gamma scale
print("***** USING rbf AND GAMMA=SCALE(set C: 1.0) ******")
rbfgas = SVC(C= 1.0, kernel= "rbf", gamma= "scale")
rbfgas.fit(X_train,y_train)
print("Score for rbf and gamma=scale: ", rbfgas.score(X_test, y_test))
print("predict digits.data[[607]]: ", rbfgas.predict(digits.data[[607]]))

y_predicted = rbfgas.predict(X_test)
print("y_predicted: \n", y_predicted)


cm = confusion_matrix(y_test, y_predicted)
print("CONFUISON MATRIX: \n", cm)

***** USING rbf AND GAMMA=SCALE(set C: 1.0) ******
Score for rbf and gamma=scale:  0.9944444444444445
predict digits.data[[607]]:  [3]
y_predicted: 
 [3 9 4 8 4 8 9 1 3 9 0 2 8 8 7 0 9 2 7 4 2 6 4 3 5 8 8 2 1 8 7 7 8 7 0 6 1
 8 0 5 8 6 7 4 2 2 9 2 1 6 5 7 3 4 1 0 0 4 5 3 2 0 9 5 4 4 9 8 1 2 8 1 8 6
 5 5 9 0 1 3 1 3 8 1 7 8 1 8 3 0 4 0 6 2 4 3 2 6 6 5 4 2 2 4 6 6 9 5 1 0 6
 4 3 0 5 7 5 6 8 4 6 4 7 6 7 4 2 9 2 4 9 8 9 7 4 0 3 9 3 9 0 6 2 9 5 3 3 9
 9 1 5 8 6 5 3 4 4 9 7 5 1 5 9 0 8 6 9 9 6 3 2 9 0 9 2 3 7 6 4 1 1 7 9 4 7
 0 9 4 9 8 2 2 5 3 9 4 9 5 7 1 1 7 8 6 1 1 6 9 9 3 1 2 2 4 6 1 2 8 1 7 3 0
 3 6 3 5 1 3 0 6 3 2 1 5 1 0 4 8 5 5 4 3 7 4 7 9 2 9 9 7 6 2 7 0 2 2 3 7 0
 5 5 6 6 5 5 5 7 3 3 0 3 3 4 5 5 8 6 7 9 6 7 4 4 3 8 2 1 9 6 3 6 7 3 4 2 8
 5 0 3 0 2 2 0 2 6 5 5 2 0 2 1 5 4 3 1 1 1 3 9 6 9 9 5 4 4 4 7 7 9 1 9 5 1
 3 7 0 4 1 9 2 4 2 0 8 8 0 6 9 3 9 6 7 0 5 9 4 0 9 2 4]
CONFUISON MATRIX: 
 [[32  0  0  0  0  0  0  0  0  0]
 [ 0 33  0  0  0  0  0  0  0  0]
 [ 0  0 38  0  0  0  0  0  0  0]


rbf with gamma= scale best parameter with 99.44% accuracy